File: src/main/scala/cccs/fluxcapacitor/tag/CachableTag.scala
package cccs.fluxcapacitor.tag

import cccs.fluxcapacitor.RuleConf
import com.google.common.hash.BloomFilter
import org.apache.log4j.Logger
import org.apache.spark.sql.Row
import cccs.fluxcapacitor.TagCache

abstract class CachableTag(
    val tagName: String,
    val ruleConf: RuleConf,
    val tagCache: TagCache
) extends Tag {

  protected def evaluate(): Boolean

  protected def makeBloomPutKey(): String

  protected def makeBloomGetKey(): String

  val baseBloomKey = ruleConf.rulename + "." + tagName
  var bloomGetKey: String = ""
  var bloomPutKey: String = ""
  var value: Boolean = false
  var row: Option[Row] = None
  var evaluated: Boolean = false

  override def toString(): String = {
    if (evaluated) {
      s"(tagName: $tagName, bloomGetKey: $bloomGetKey, bloomPutKey: $bloomPutKey, value: $value, isCached: ${isCached()})"
    } else {
      s"(tagName: $tagName)"
    }
  }

  override def evaluateRow(row: Row): Boolean = {
    this.row = Option(row)
    bloomGetKey = makeBloomGetKey()
    bloomPutKey = makeBloomPutKey()
    evaluated = true
    evaluate()
  }

  private def getColumnValue(colname: String): String = {
    if (colname.isEmpty()) {
      ""
    } else {
      var i = row.get.fieldIndex(colname)
      if (row.get.isNullAt(i)) {
        ""
      } else {
        row.get.getString(i)
      }
    }
  }

  protected def makeBloomKey(colname: String = ""): String = {
    if (colname.isEmpty()) {
      baseBloomKey
    } else {
      var suffix = getColumnValue(colname)
      if (log.isTraceEnabled)
        log.trace(s"suffix [$suffix] retrieved for column [$colname]")
      baseBloomKey + "." + suffix
    }
  }

  def storeInCache(): Unit = {
    if(log.isTraceEnabled()) log.trace(s"storing key: $bloomPutKey in bloom")
    tagCache.put(bloomPutKey)
  }

  def isCached(): Boolean = tagCache.get(bloomGetKey)

  override def setCurrentValue(newValue: Boolean): Tag = {
    bloomGetKey = ""
    bloomPutKey = ""
    evaluated = false
    this.value = newValue
    this
  }

  def currentValue(): Boolean = value

  protected def mergeCachedWithCurrent() = {
    this.value = currentValue() | isCached()
  }

  protected def pushOrPullFromCache() = {
    if (currentValue()) {
      // put "this" tag in the bloom
      if (log.isTraceEnabled())
        log.trace("the tag was raised in this row, store tag in bloom")
      storeInCache()
    } else {
      // get "this" tag from the bloom
      if (log.isTraceEnabled())
        log.trace("current value is false, cached value is: " + isCached())
      mergeCachedWithCurrent()
      if (log.isTraceEnabled()) log.trace("merged value is " + currentValue())
    }
  }

}

File: src/main/java/cccs/oldversion/tag/CachableTag.java
package cccs.oldversion.tag;

import org.apache.log4j.Logger;
import org.apache.spark.sql.Row;

import com.google.common.hash.BloomFilter;

import cccs.oldversion.RulesConf.RuleConf;

public abstract class CachableTag extends Tag {
  static Logger log = Logger.getLogger(CachableTag.class.getName());
  private String bloomGetKey;
  private String bloomPutKey;
  private boolean value;
  protected Row row;
  private final BloomFilter<CharSequence> bloom;
  protected final RuleConf ruleConf;
  protected final String tagname;

  public CachableTag(String tagname, RuleConf ruleConf, BloomFilter<CharSequence> bloom) {
    this.tagname = tagname;
    this.ruleConf = ruleConf;
    this.bloom = bloom;
  }

  @Override
  public final boolean evaluateRow(Row row) {
    this.row = row;
    bloomGetKey = makeBloomGetKey();
    bloomPutKey = makeBloomPutKey();
    return evaluate();
  }

  protected abstract boolean evaluate();
  
  protected abstract String makeBloomPutKey();

  protected abstract String makeBloomGetKey();

  private static String getColumnValue(String colname, Row row) {
		if (!colname.isEmpty()) {
			int i = row.fieldIndex(colname);
			if (row.isNullAt(i)) {
				return "";
			}
			return row.getString(i);
		} else {
			return "";
		}
	}

	protected static String makeBloomKey(String colname, String ruleName, String tagName, Row row) {
		String prefixValue = getColumnValue(colname, row);
    log.debug("row value retrieved for column: " + colname + " is: [" + prefixValue + "]");
		return ruleName + "." + prefixValue + "." + tagName;
	}

  public void storeInCache() {
    log.debug("storing key: " + bloomPutKey + " in bloom");
    bloom.put(bloomPutKey);
  }

  protected boolean isCached() {
    boolean ret = bloom.mightContain(bloomGetKey);
    if (ret) {
      log.debug("key: " + bloomGetKey + " found in bloom");
    } else {
      log.debug("key: " + bloomGetKey + " not found in bloom");
    }
    return ret;
  }

  @Override
  public void setCurrentValue(boolean value){
    reset(value);
  }

  private void reset(boolean value){
    this.value = value;
    this.bloomGetKey = null;
    this.bloomPutKey = null;
  }


  public boolean currentValue() {
    return value;
  }

  protected void mergeCachedWithCurrent() {
    value = currentValue() | isCached();
  }

  protected void pushOrPullFromCache() {
    if (currentValue()) {
      // put "this" tag in the bloom
      log.debug("the tag was raised in this row, store tag in bloom");
      storeInCache();
    } else {
      // get "this" tag from the bloom
      log.debug("the tag is not raised in this row, cached value is: " + isCached());
      mergeCachedWithCurrent();
      log.debug("merged value is " + currentValue());
    }
  }

}